godot-core/src/builtin/arrays.rs

```diff
@@ -112,12 +120,15 @@ impl Array {
         Ok(vec)
     }
 
-    /// Implements iteration over an `Array` by reference, but returns (cheap) copies of the
-    /// `Variant`s in the array.
+    /// Returns an iterator over the `Array` by reference. Instead of references to elements as you
+    /// might expect, the iterator returns a (cheap, shallow) copy of each element.
+    ///
+    /// Notice that it's possible to modify the `Array` through another reference while iterating
+    /// over it. This will not result in unsoundness or crashes, but will cause the iterator to
+    /// behave in an unspecified way.
     pub fn iter(&self) -> ArrayIterator<'_> {
```

Comment:
Would you leave the name `iter()`, or something like `open_iter()`? Not that I particularly like the name or verbosity, but the likelyhood that people read docs on unfamiliar names is much higher, while they associate certain expectations to familiar ones. `iter()` is one the things that works exactly the same way everywhere in Rust, but not here 🤔 

Some brainstorming from the Discord: non-unique/shared/common/open/exposed/susceptible/prone/...
"Open" is quite short among those.

Reply:
If the name must be changed, I would prefer `iter_shared` (not `shared_iter` because autocomplete is more helpful if the name starts with `iter`).

But what about the `IntoIterator` impl? Creating a new trait for this, like `IntoSharedIterator`, would limit its usefulness with existing APIs that take `IntoIterator`.

Reply:
Hmm, good points. Is `IntoIterator` even a good idea under these circumstances?

`for elem in &array {...}` is nice syntax, but it could be _even more_ surprising to learn that the shared-ref does not prevent any mutations of the underlying `array`. Maybe it's really better to be a bit more explicit here.

Is auto-complete truly an issue? This API is super-small.
In CLion, `shared_iter()` is suggested on top even if I just type "it":
![grafik](https://user-images.githubusercontent.com/708488/215169972-fb6768eb-3bd7-4e2a-afb9-ca4825b5b08f.png)

_(It's more annoying that the `#[doc(hidden)]` methods show up, but at least they're on the bottom)._

Reply:
To be honest I never bothered to set up autocomplete in vim :) But `iter_shared` is more discoverable in HTML docs too, and it follows the pattern of `iter_mut` and `BinaryHeap::into_iter_sorted` from `std`.

`IntoIterator` is nice because it appears in other APIs, such as `Extend`. But since `Iterator` also implements `IntoIterator`, no functionality is lost by removing the `IntoIterator` impl on `Array`; you just need to pass `array.iter_shared()` instead of `array` itself.

Reply:
> But `iter_shared` is more discoverable in HTML docs too, and it follows the pattern of `iter_mut` and `BinaryHeap::into_iter_sorted` from `std`.

Good point, let's go with `iter_shared` then.

Thanks for removing `IntoIterator`!